torch/csrc/jit/passes/to_batch.cpp

```diff
@@ -1,63 +1,290 @@
+  if (last_dot_pos != std::string::npos && last_dot_pos + 1 != name.size()) {
+    if (name.find_first_not_of("0123456789", last_dot_pos + 1) == std::string::npos) {
+      auto suffix = name.substr(last_dot_pos + 1);
+      std::string name_base = name.substr(0, last_dot_pos);
+      return {name_base, suffix};
+    }
+  }
+  return {name};
+}
+
+// replace aten operator node with BatchTensor operator graph
+void ToBatch::visitAten(Node* n, Block* block, Block* res_block, std::unordered_map<std::string, Value*>& var_map){
+  if(n->outputs().size() > 1){
+    throw std::runtime_error("Cannot process multiple assignment");
+  }
+  auto batch_graph = batch_operator_table.at(n->kind().toUnqualString());
+  std::vector<Value*> new_inputs;
+  for(Value *input : n->inputs()){
+    if(batch_map.find(input) != batch_map.end()){
+      auto new_input = batch_map.at(input);
+      new_inputs.insert(new_inputs.end(), new_input.begin(), new_input.end());
+    }
+    else{
+      throw std::runtime_error("NYI: non-tensor input for aten operator is not supported yet");
+    }
+  }
+  auto outputs = script::inlineCallTo(*res_block->owningGraph(), *batch_graph, new_inputs);
+
+  // do update on assignment
+  auto name_base = get_name(n->output()->uniqueName())[0];
+  if(var_map.find(name_base) != var_map.end()){
+    std::vector<Value*> inputs(batch_map.at(var_map.at(name_base)));
+    inputs.insert(inputs.end(), outputs.begin(), outputs.end());
+    outputs = script::inlineCallTo(*res_block->owningGraph(), *batch_operator_table.at("update"), inputs);
+  }
+  // Assume all outputs from inlined operator implementation are in the triple form.
+  for(size_t i = 0; i < n->outputs().size(); i++){
+    auto output = n->outputs()[i];
+    batch_map[output] = std::vector<Value*>(outputs.begin() + i * 3, outputs.begin() + i * 3 + 3);
+    if(output->hasUniqueName()){
+      var_map[get_name(output->uniqueName())[0]] = output;
+    }
+  }
+}
+
+// clone prim::Constant to new graph
+void ToBatch::visitConstant(Node* n, Block* block, Block* res_block){
+  auto res_graph = res_block->owningGraph();
+  auto* r_node = res_graph->createClone(n, rn_fn);
+  r_node->setStage(n->stage());
+  res_graph->appendNode(r_node);
+  rn_env[n->output()] = r_node->output();
+}
+
+// change return tensor to {data, mask, dims}
+void ToBatch::visitNumToTensor(Node* n, Block* block, Block* res_block){
+  auto res_graph = res_block->owningGraph();
+  auto* r_node = res_graph->createClone(n, rn_fn);
+  r_node->setStage(n->stage());
+  res_graph->appendNode(r_node);
+  auto outputs = script::inlineCallTo(*res_block->owningGraph(), *batch_operator_table.at("batch_from_scalar_tensor"), r_node->outputs());
+  batch_map[n->output()] = outputs;
+}
+
+// elif is not supported
+// assume every variable assigned in an if statement is already defined before
+void ToBatch::visitIf(Node* n, Block* block, Block* res_block, std::unordered_map<std::string, Value*>& var_map){
```

Comment:
For both `If` and `While`, can you add a block comment above each of these functions and provide before/after examples for the graph when it comes in and the graph when it goes through here? There are lots of subtle details and it'd be good to be able to visualize what's going on with examples